File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/currency/service/BYRExchangeRateChartService.java
package com.github.bumblebee.command.currency.service;

import com.github.bumblebee.command.currency.config.CurrencyChartConfig;
import com.github.bumblebee.command.currency.config.CurrencyCommandLocalization;
import com.github.bumblebee.command.currency.domain.DailyExchangeRate;
import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartUtilities;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.labels.StandardCategoryItemLabelGenerator;
import org.jfree.chart.plot.CategoryPlot;
import org.jfree.chart.plot.PlotOrientation;
import org.jfree.chart.renderer.category.CategoryItemRenderer;
import org.jfree.data.category.CategoryDataset;
import org.jfree.data.category.DefaultCategoryDataset;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class BYRExchangeRateChartService {

    private static final double MARGIN_PERCENT = 0.1d;
    private static final int DEFAULT_CHART_WIDTH = 600;
    private static final int DEFAULT_CHART_HEIGHT = 500;
    private static final int CATEGORY_WIDTH = 70;

    private final CurrencyChartConfig currencyChartConfig;

    @Autowired
    public BYRExchangeRateChartService(CurrencyChartConfig currencyChartConfig) {
        this.currencyChartConfig = currencyChartConfig;
    }

    public byte[] createChartImage(List<DailyExchangeRate> rates, boolean detailed, LocalDate from, LocalDate to) throws IOException {

        rates.sort(Comparator.comparing(DailyExchangeRate::getDate));

        JFreeChart chart;
        if (detailed) {
            chart = createBarChart(rates);
        } else {
            chart = createLineChart(rates);
        }
        BufferedImage image = chart.createBufferedImage(getWidth(chart), DEFAULT_CHART_HEIGHT);
        return ChartUtilities.encodeAsPNG(image);
    }

    private JFreeChart createLineChart(List<DailyExchangeRate> rates) {
        CurrencyCommandLocalization localization = currencyChartConfig.getLocalization();
        JFreeChart chart = ChartFactory.createLineChart(
                localization.getChartTitle(),
                localization.getChartDateAxis(),
                localization.getChartRateAxis(),
                createDataset(rates, true),
                PlotOrientation.VERTICAL, true, true, false);

        CategoryPlot categoryPlot = chart.getCategoryPlot();
        double minimumRate = getMinimumRate(rates);
        categoryPlot.getRangeAxis().setLowerBound(minimumRate - minimumRate * 0.01);

        return chart;
    }

    private JFreeChart createBarChart(List<DailyExchangeRate> rates) {
        CurrencyCommandLocalization localization = currencyChartConfig.getLocalization();
        JFreeChart chart = ChartFactory.createBarChart(
                localization.getChartTitle(),
                localization.getChartDateAxis(),
                localization.getChartRateAxis(),
                createDataset(rates, false),
                PlotOrientation.VERTICAL, true, true, false);

        CategoryPlot categoryPlot = chart.getCategoryPlot();
        CategoryItemRenderer renderer = categoryPlot.getRenderer();
        // draw values on bars
        renderer.setBaseItemLabelGenerator(new StandardCategoryItemLabelGenerator());
        renderer.setBaseItemLabelsVisible(true);
        categoryPlot.setRenderer(renderer);

        // top margin
        double maxRate = getMaximumRate(rates);
        categoryPlot.getRangeAxis().setUpperBound(maxRate + maxRate * MARGIN_PERCENT);

        return chart;
    }

    private CategoryDataset createDataset(List<DailyExchangeRate> rates, boolean showMonth) {
        final DefaultCategoryDataset data = new DefaultCategoryDataset();
        final String format = (showMonth) ? "MMM yy" : "d.MM";
        final SimpleDateFormat formatter = new SimpleDateFormat(format, Locale.ENGLISH);

        rates.forEach(rate -> data.addValue(
                rate.getRate(),
                rate.getCurrency(),
                formatter.format(rate.getDate())
        ));
        return data;
    }

    private double getMaximumRate(List<DailyExchangeRate> rates) {
        return rates.stream()
                .mapToDouble(DailyExchangeRate::getRate)
                .max()
                .getAsDouble();
    }

    private double getMinimumRate(List<DailyExchangeRate> rates) {
        return rates.stream()
                .mapToDouble(DailyExchangeRate::getRate)
                .min()
                .getAsDouble();
    }

    private int getWidth(JFreeChart chart) {
        int autoWidth = chart.getCategoryPlot().getCategories().size() * CATEGORY_WIDTH;
        return (autoWidth > DEFAULT_CHART_WIDTH) ? autoWidth : DEFAULT_CHART_WIDTH;
    }

    public List<DailyExchangeRate> reduceToAverageByMonth(List<DailyExchangeRate> rates) {

        SimpleDateFormat format = new SimpleDateFormat("yyyy.MM");

        Map<String, Map<String, List<DailyExchangeRate>>> grouping = rates.stream()
                .collect(Collectors.groupingBy(DailyExchangeRate::getCurrency,
                        Collectors.groupingBy(rate -> format.format(rate.getDate()))));

        List<DailyExchangeRate> reduced = new ArrayList<>();
        grouping.forEach((currency, groupedRates) ->
                groupedRates.values().forEach(monthRates -> {
                    double monthAvg = monthRates.stream()
                            .mapToDouble(DailyExchangeRate::getRate)
                            .average()
                            .getAsDouble();
                    DailyExchangeRate sampleRate = monthRates.get(0);
                    reduced.add(new DailyExchangeRate(sampleRate.getDate(), sampleRate.getCurrency(), monthAvg));
                })
        );
        return reduced;
    }
}

File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/currency/service/BYRExchangeRateChartService.kt
package com.github.bumblebee.command.currency.service

import com.github.bumblebee.command.currency.config.CurrencyChartConfig
import com.github.bumblebee.command.currency.domain.DailyExchangeRate
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartUtilities
import org.jfree.chart.JFreeChart
import org.jfree.chart.labels.StandardCategoryItemLabelGenerator
import org.jfree.chart.plot.PlotOrientation
import org.jfree.data.category.CategoryDataset
import org.jfree.data.category.DefaultCategoryDataset
import org.springframework.stereotype.Service
import java.io.IOException
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.util.*

@Service
class BYRExchangeRateChartService(private val currencyChartConfig: CurrencyChartConfig) {

    @Throws(IOException::class)
    fun createChartImage(rates: List<DailyExchangeRate>, detailed: Boolean, from: LocalDate, to: LocalDate): ByteArray {
        val sortedRates = rates.sortedBy { it.date }

        val chart = if (detailed)
            createBarChart(sortedRates)
        else
            createLineChart(sortedRates)
        val image = chart.createBufferedImage(getWidth(chart), DEFAULT_CHART_HEIGHT)
        return ChartUtilities.encodeAsPNG(image)
    }

    private fun createLineChart(rates: List<DailyExchangeRate>): JFreeChart {
        val localization = currencyChartConfig.localization
        val chart = ChartFactory.createLineChart(
                localization.chartTitle,
                localization.chartDateAxis,
                localization.chartRateAxis,
                createDataset(rates, true),
                PlotOrientation.VERTICAL, true, true, false)

        val minimumRate = getMinimumRate(rates)
        chart.categoryPlot.rangeAxis.lowerBound = minimumRate - minimumRate * 0.01

        return chart
    }

    private fun createBarChart(rates: List<DailyExchangeRate>): JFreeChart {
        val localization = currencyChartConfig.localization
        val chart = ChartFactory.createBarChart(
                localization.chartTitle,
                localization.chartDateAxis,
                localization.chartRateAxis,
                createDataset(rates, false),
                PlotOrientation.VERTICAL, true, true, false)

        val categoryPlot = chart.categoryPlot
        val renderer = categoryPlot.renderer
        // draw values on bars
        renderer.baseItemLabelGenerator = StandardCategoryItemLabelGenerator()
        renderer.setBaseItemLabelsVisible(true)
        categoryPlot.renderer = renderer

        // top margin
        val maxRate = getMaximumRate(rates)
        categoryPlot.rangeAxis.upperBound = maxRate + maxRate * MARGIN_PERCENT

        return chart
    }

    private fun createDataset(rates: List<DailyExchangeRate>, showMonth: Boolean): CategoryDataset {
        val data = DefaultCategoryDataset()
        val format = if (showMonth) "MMM yy" else "d.MM"
        val formatter = SimpleDateFormat(format, Locale.ENGLISH)

        rates.forEach { rate ->
            data.addValue(
                    rate.rate,
                    rate.currency,
                    formatter.format(rate.date)
            )
        }
        return data
    }

    private fun getMaximumRate(rates: List<DailyExchangeRate>): Double {
        return rates.maxBy { it.rate }?.rate ?: 0.0
    }

    private fun getMinimumRate(rates: List<DailyExchangeRate>): Double {
        return rates.minBy { it.rate }?.rate ?: 0.0
    }

    private fun getWidth(chart: JFreeChart): Int {
        val autoWidth = chart.categoryPlot.categories.size * CATEGORY_WIDTH
        return if (autoWidth > DEFAULT_CHART_WIDTH) autoWidth else DEFAULT_CHART_WIDTH
    }

    fun reduceToAverageByMonth(rates: List<DailyExchangeRate>): List<DailyExchangeRate> {
        val format = SimpleDateFormat("yyyy.MM")

        val grouping = rates.groupBy { it.currency }
                .mapValues { (_, currencyRates) -> currencyRates.groupBy { format.format(it.date) } }

        val reduced = ArrayList<DailyExchangeRate>()
        grouping.forEach { _, groupedRates ->
            groupedRates.values.forEach { monthRates ->
                val monthAvg = monthRates.map { it.rate }.average()
                val sampleRate = monthRates[0]
                reduced.add(DailyExchangeRate(sampleRate.date, sampleRate.currency, monthAvg))
            }
        }
        return reduced
    }

    companion object {
        private val MARGIN_PERCENT = 0.1
        private val DEFAULT_CHART_WIDTH = 600
        private val DEFAULT_CHART_HEIGHT = 500
        private val CATEGORY_WIDTH = 70
    }
}
